### ISSUE TITLE
[RFC]: Model customized KVCache Planning

### ISSUE THREAD

**Issue by MengqingCao** [vllm-project/vllm#44276]
### Motivation.
> plz refer to https://docs.google.com/document/d/1UwD65kGqJQOoLj1i_0xC-dvv0ojz_nANxkJ95u1QFEY/edit?usp=sharing to read the details
### Current KVCache planning
Currently all the KVCache planning strategies are covered by the `kv_cache_utils.py`. It is treated as a global feature for each model. The core steps including:

- merge the kv cache specs from each worker
- group the kv cache specs
- maybe override num_blocks
- auto-fix max model len
- create `KVCacheTensor` and then `KVCacheConfig` according to the group

The most important but also heavy work is grouping the kv cache specs. There’re 5 branches in the general scenarios:
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-c27a1b20-7fff-2ed4-46d1-36bf2c5bf9c9"><div dir="ltr" style="margin-left:5.25pt;" align="left">
number of kvcache | grouping method | number of groups
-- | -- | --
/ | use unified kv cache specs | 1
0 | no grouping | 0
1 | uniform kv cache specs | 1
usually 2 (sizes differ between layers) | uniform type kv cache specs | 1
usually 2 (types and sizes differ) | hybrid kv cache grouping: align the page size between different type of specs; determine a group size for sharing the same buffer; group the kv cache specs by `KVCacheGroupSepcs` | num_layers/group_size

</div><br /></b>

With the hybrid attention models growing, HMA plays an important role in kv cahce planning. Taking advantage of HMA, when the attention types in the model are within 2, we could handle it with a uniform logic. The models such as Gemma3, Qwen3.5, Bailing, etc. all benefit from HMA.

### Challenge
But after the coming of DeepSeek-V4, there are 7 types of cache we need to take into account. 

<img width="383" height="341" alt="Image" src="https://github.com/user-attachments/assets/67605620-5a9a-4f8c-8bab-375d9327fec2" />

> the figure comes from [https://vllm.ai/blog/2026-04-24-deepseek-v4](https://vllm.ai/blog/2026-04-24-deepseek-v4)

It is a huge challenge to use a uniform grouping mechanism and make great use of memory, thus vLLM offers a specific way for DeepSeek-V4 for a more effective kvcache planning. And the main differences:

- use `UniformTypeKVCacheSpecs` to pack the specs that would be grouped into 1 group
- align page sizes between different `UniformTypeKVCacheSpecs`, following the mechanism that `MLAAttentionSpec` plays role as a standard size, and align the `SlidingWindowMLASpec` with the nearest page size of `MLAAttentionSpec`
- creating `num_layer_tuples` tensors, instead of 1.

These differences complicate the logic of core key-value caching planning, but only works for DeepSeek-V4. This not only makes the code architecture unreasonable, but the situation will be even worse if a new model with a more complex kvcache design is introduced.

### Proposed Change.

To address the above challenge, this rfc introduces a model-customized kvcache planning feature into vLLM. We moved the kvcache planning job from core to models, allowing them to handle it themselves. 

### KVCacheConfig
Generate `KVCacheConfig` through `KVCachePlanner`, and dispatch the planner according to a new config: `ModelConfig.kv_cache_planner_cls` in core. 

If the `model_config.kv_cache_planner_cls` is `None`, it means no model-specific planner is implemented. This meets the scenario of most models (including pure single type attention or hybrid attentions within 2 types), whose kvcache is simple, will keep the kvcache planning method as before. And the planning logic of them will be centralized into `DefaultKVCachePlanner`. 

<img width="1045" height="455" alt="Image" src="https://github.com/user-attachments/assets/bd8f3b0d-1076-46b6-bb3c-db71a7b84029" />

Otherwise if the `model_config.kv_cache_planner_cls` is not None, it means there is a customized planner set for the model, for example:

```python
class DeepseekV4ForCausalLM(nn.Module):
    model_cls = DeepseekV4Model
    kv_cache_planner_cls = (
        "reference.path.to.DeepseekV4KVCachePlanner"
    )
    def __init__(self, *, vllm_config: VllmConfig, prefix: str = ""):
       ...

```

The KVCachePlanner should handle the following responsibilities:

1. kvcahce spec refreshing (maybe plan `block_size` or `pad_size`, etc.)
2. kvcache specs grouping
3. max model len auto-fixing
4. `KVCacheTensor` and `KVCacheConfig` creation

**require discussion:**
the APIs should be included in an abstract class of KVCachePlanner. and also only these functions will be marked as public methods.

- `get_kv_cache_groups`: keep this api as we need this in when profiling cudagraph mem in `_init_minimal_kv_cache_for_profiling`
- `get_kv_cache_configs_from_groups`: ditto
- `get_kv_cache_configs`: this is at a high enough level to meet the custom requirements of the key-value caching plan. 
- `get_max_model_len:` we expose this api for querying max model length the llm could serve.

**the final file tree:**
The kvcache planning is proposed to put with the modeling files, and maybe the default kvcache planner could be separated into 2 types, with and without HMA? This may  need more discussion.

> Note
> There will be only one kvcache planner for `normal_hybrid_models` and `single_attention_type_models` in the first version.

```
|- model_executor
    |- models
       |- complex_hybrid_models
          |- deepseek_v4
              |- kvcahce_planning
		       |- kv_cache_planner
		       |- kv_cache_spec
              |- modeling
              |- mtp_modeling
       |- normal_hybrid_models
           |- kv_cahce_planner
           |- Qwen3.5
           |- Gemma3
           |- ...
       |- single_attention_type_models
           |- kv_cahce_planner
           |- Qwen3
           |- deepseek_v3_2
           |- ...

```

**An example for DeepSeek V4 KVCachePlanner**

```python
class DeepseekV4KVCachePlanner:
    """
Handle the whole kvcache planning of DeepSeek V4, including the process of kvcache spec, grouping, `KVCacheTensor` definition and `KVCacheConfig` creation.
    """

   def __init__(self, vllm_config: VllmConfig):
        """Create a planner bound to the engine cache configuration."""

        self.vllm_config = vllm_config
        # main_block_size refers to the block_size of MLAAttentionSpec
        self.main_block_size = vllm_config.cache_config.block_size
        self.supported_compress_ratios = (
            set(vllm_config.model_config.hf_config.compress_ratios) | {1}
        )

    def _post_process_kv_cache_specs(self, kv_cache_spec: dict[str, KVCacheSpec]) -> dict[str, KVCacheSpec]:
        """Return KV cache specs with DeepSeek V4 block/page sizes finalized."""

        spec_infos = self._classify_specs(kv_cache_spec)
        canonical_buckets = self._build_canonical_buckets(spec_infos)
        inferred_layouts = self._infer_sliding_win_mla_spec_layouts(
            spec_infos, canonical_buckets
        )
        finalized = self._finalize_specs(
            kv_cache_spec, spec_infos, canonical_buckets, inferred_layouts
        )
        return finalized

    def get_kv_cache_groups(
        self, kv_cache_specs: dict[str, KVCacheSpec]
    ) -> list[KVCacheGroupSpec]:
        """Build DeepSeek V4 KV cache groups from finalized specs."""

        grouped_specs = self._group_and_unify_specs(kv_cache_specs)
        kv_cache_groups = self._get_kv_cache_groups_from_uniform_groups(grouped_specs)
        self._annotate_eagle_groups(kv_cache_specs, kv_cache_groups)
        return kv_cache_groups
       def get_kv_cache_configs(
        self,
        kv_cache_specs: list[dict[str, KVCacheSpec]],
        available_memory: list[int],
    ) -> list[KVCacheConfig]:
        """
        Generates the KV cache configurations for a model.

        1. Merges the KV cache specs of all workers to get the KVCacheSpecs for
           the whole model.
        2. Generates the KV cache groups based on the layer ratio of the whole
           model. This also handles spec unification for hybrid models.
        3. Handles auto-fit max_model_len and memory checks using per-worker
           projected groups to account for PP sharding.
        4. Generates the KV cache configs for each worker based on the KV cache
           grouping strategy.
        5. Changes the num_blocks of each worker to the smallest among all
           workers and shrinks tensor sizes proportionally.
        """


        ...
   
def get_kv_cache_config_from_groups(
        self,
        kv_cache_groups: list[KVCacheGroupSpec],
        available_memory: int,
    	) -> KVCacheConfig:
		...
```

### CC List.

@heheda12345 @WoosukKwon @ivanium @ZJY0516 @zhewenl 

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by mengweihuang** [comment]
The models layer is currently working on decoupling from the hardware（ #42770 ）.  Since this RFC also involves refactoring the models layer, it looks like it might make the models layer more complicated.  How should we align these changes?

**Comment by LucasWilkinson** [comment]
Thanks for the RFC! I am generally in favor of giving the model more hooks to override behavior like this. However I think we should wait for https://github.com/vllm-project/vllm/issues/42449 to complete and then re-evaluate this given that plan to remove `UniformTypeKVCacheSpecs` as part of the "2. Backends Declare Manager + Tensor Spec" section. Any feed back is very much welcome!

**Comment by MengqingCao** [comment]
> The models layer is currently working on decoupling from the hardware（ [#42770](https://github.com/vllm-project/vllm/issues/42770) ）.  Since this RFC also involves refactoring the models layer, it looks like it might make the models layer more complicated.  How should we align these changes?

Hi, thanks for your reply! In fact, I think `KVCachePlanner` is a very independent feature. It only introduces a single configuration parameter `model_config.kv_cache_planner_cls` to the model, yet it achieves customization of kvcache planning. 

Furthermore, if similar kvcache planning models emerge later, I believe it's necessary to abstract and unify them so that a group of models can share the same kvcache planning strategy.

**Comment by MengqingCao** [comment]
> Thanks for the RFC! I am generally in favor of giving the model more hooks to override behavior like this. However I think we should wait for [#42449](https://github.com/vllm-project/vllm/issues/42449) to complete and then re-evaluate this given that plan to remove `UniformTypeKVCacheSpecs` as part of the "2. Backends Declare Manager + Tensor Spec" section. Any feed back is very much welcome!

Thanks for this suggestion! I'm okay to do this job after #42449, but I want to confirm more details on it.

I've read #42449 and I think it solve a lot of complexity effectively, but maybe I have some misunderstanding on that. plz point it out if anything is wrong:

The "2. Backend Declaration Manager + Tensor Specification" section of #42449 attempts to integrate the existing `get_kv_cache_specs` and `SingleTypeKVCacheManager` into a single attention backend, allowing kvcache-related parameters (such as sliding window size), kvcache manager cls, and KVCache physical memory layout to be directly defined in the attn backend.

However, regarding the **kvcache grouping logic**, including the padding for a uniform page size, where and how do you plan to handle this? I only noticed that you wrote it on the worker side, but I don't know the specific details. Because it needs to collect kvcache-related information from all layers (attention backend), the subsequent grouping and planning is currently a very heavy workload. Delegating all of this to the attention backend seems to transfer the planning complexity from the original core to the attention backend.

Therefore, could I conclude that #42449 ideally involves centralizing the **kvcache parameters, manager cls, tensor layout**, etc., in the attention backend. while the original **HMA grouping logic** can be handled by the `KVCachePlanner` I proposed here?
